src/main/java/hudson/plugins/ec2/EC2Cloud.java

```diff
@@ -165,14 +162,20 @@
 
     private transient volatile AmazonEC2 connection;
 
-    protected EC2Cloud(String id, boolean useInstanceProfileForCredentials, String credentialsId, String privateKey,
-            String instanceCapStr, List<? extends SlaveTemplate> templates, String roleArn, String roleSessionName) {
+    protected EC2Cloud(String id, boolean useInstanceProfileForCredentials, String credentialsId, String privateKey, String sshKeysCredentialsId,
+                       String instanceCapStr, List<? extends SlaveTemplate> templates, String roleArn, String roleSessionName) {
         super(id);
         this.useInstanceProfileForCredentials = useInstanceProfileForCredentials;
         this.roleArn = roleArn;
         this.roleSessionName = roleSessionName;
         this.credentialsId = credentialsId;
-        this.privateKey = new EC2PrivateKey(privateKey);
+        this.sshKeysCredentialsId = sshKeysCredentialsId;
+
+        if (this.sshKeysCredentialsId == null && ( this.privateKey != null || privateKey != null)){
+            migratePrivateSshKeyToCredential(this.privateKey != null ? this.privateKey.getPrivateKey() : privateKey);
```

Comment:
A migration call in the constructor makes no sense. This should be in `readResolve`. This was probably never tested.